pandas_dev_flaker/_plugins_tree/disallow_argument_types.py

```diff
@@ -0,0 +1,47 @@
+import ast
+from typing import Iterator, Tuple
+
+from pandas_dev_flaker._data_tree import State, register
+
+MSG = "PDF026 found union between Series and AnyArrayLike in type hint"
+
+
+@register(ast.FunctionDef)
+def visit_arg(
+    state: State,
+    node: ast.FunctionDef,
+    parent: ast.AST,
+) -> Iterator[Tuple[int, int, str]]:
+    arguments = node.args.args
+    for arg in arguments:
+        if isinstance(arg.annotation, ast.BinOp) and parse_annotation(
+            arg.annotation,
+        ):
+            yield arg.lineno, arg.col_offset, MSG
+
+
+def parse_annotation(node: ast.BinOp) -> bool:
+    series, any_array_like = "Series", "AnyArrayLike"
+    is_series, is_array_like = False, False
+
+    while True:
+        if isinstance(node.right, ast.Name):
+            if node.right.id == series:
+                is_series = True
+            if node.right.id == any_array_like:
+                is_array_like = True
+
+        if isinstance(node.left, ast.Name):
+            if node.left.id == series:
+                is_series = True
+            if node.left.id == any_array_like:
+                is_array_like = True
+
+        if isinstance(node.left, ast.BinOp):
+            node = node.left
+        elif isinstance(node.left, ast.Name):
+            break
+        else:
+            raise AssertionError
```

Comment:
Try running this on `pandas`, this'll be reached
```
$ flake8 ../pandas-marco/pandas --select PDF026 -j 1
Traceback (most recent call last):
  File "/home/marco/pandas-style-guide/venv/bin/flake8", line 8, in <module>
    sys.exit(main())
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/main/application.py", line 363, in run
    self._run(argv)
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/main/application.py", line 351, in _run
    self.run_checks()
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/main/application.py", line 264, in run_checks
    self.file_checker_manager.run()
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/checker.py", line 323, in run
    self.run_serial()
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/checker.py", line 307, in run_serial
    checker.run_checks()
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/checker.py", line 589, in run_checks
    self.run_ast_checks()
  File "/home/marco/pandas-style-guide/venv/lib/python3.8/site-packages/flake8/checker.py", line 494, in run_ast_checks
    for (line_number, offset, text, _) in runner:
  File "/home/marco/pandas-style-guide/pandas_dev_flaker/__main__.py", line 22, in run
    for line, col, msg in callbacks_tree:
  File "/home/marco/pandas-style-guide/pandas_dev_flaker/_data_tree.py", line 74, in visit_tree
    yield from ast_func(state, node, parent)
  File "/home/marco/pandas-style-guide/pandas_dev_flaker/_plugins_tree/disallow_argument_types.py", line 17, in visit_arg
    if isinstance(arg.annotation, ast.BinOp) and parse_annotation(
  File "/home/marco/pandas-style-guide/pandas_dev_flaker/_plugins_tree/disallow_argument_types.py", line 45, in parse_annotation
    raise AssertionError
AssertionError
```

Reply:
This is the example that breaks this:
```
def maybe_iterable_to_list(obj: Iterable[T] | T) -> Collection[T] | T:
```

Reply:
wow, wow, wow. i didn't think of this edge case